src/loaders/OmeZarrLoader.ts

```diff
@@ -325,102 +425,63 @@ class OMEZarrLoader implements IVolumeLoader {
     vol.imageInfo = {
       ...vol.imageInfo,
-      atlasTileDims: new Vector2(nrows, ncols),
+      atlasTileDims,
       volumeSize: volSizePx,
       subregionSize: regionSizePx,
-      subregionOffset: regionExtentPx.min,
+      subregionOffset: regionPx.min,
     };
     vol.updateDimensions();
 
-    const channelIndexes = vol.loadSpec.channels || Array.from({ length: numChannels }, (_val, idx) => idx);
-    // do each channel on a worker
-    for (const i of channelIndexes) {
-      const cacheQueryDims = {
-        region: regionPx,
-        time: vol.loadSpec.time,
-        scale: levelToLoad,
-      };
-      const cacheResult = this.cacheStore && this.cache?.get(this.cacheStore, i, cacheQueryDims);
-
-      if (cacheResult) {
-        vol.setChannelDataFromVolume(i, cacheResult);
-        onChannelLoaded?.(vol, i);
-      } else {
-        const worker = new Worker(new URL("../workers/FetchZarrWorker", import.meta.url));
-        worker.onmessage = (e) => {
-          const u8 = e.data.data;
-          const channel = e.data.channel;
-
-          const cacheInsertDims = {
-            region: regionPx,
-            scale: levelToLoad,
-            time: vol.loadSpec.time,
-            channel,
-          };
-          this.cacheStore && this.cache?.insert(this.cacheStore, u8, cacheInsertDims);
-
-          vol.setChannelDataFromVolume(channel, u8);
-          onChannelLoaded?.(vol, channel);
-          worker.terminate();
-        };
-        worker.onerror = (e) => {
-          alert("Error: Line " + e.lineno + " in " + e.filename + ": " + e.message);
-        };
-
-        const msg: FetchZarrMessage = {
-          url: this.url,
-          subregion: regionPx,
-          time: Math.min(vol.loadSpec.time, times),
-          channel: i,
-          path: datasetPath,
-          axesTCZYX: this.axesTCZYX,
-        };
-        worker.postMessage(msg);
-      }
-    }
+    const { numChannels } = vol.imageInfo;
+    const channelIndexes = vol.loadSpec.channels ?? Array.from({ length: numChannels }, (_, i) => i);
+
+    const channelPromises = channelIndexes.map(async (ch) => {
+      // Build slice spec
+      const { min, max } = regionPx;
+      const unorderedSpec = [vol.loadSpec.time, ch, slice(min.z, max.z), slice(min.y, max.y), slice(min.x, max.x)];
+      const specLen = getDimensionCount(this.axesTCZYX);
+      const sliceSpec: (number | Slice)[] = Array(specLen);
+
+      this.axesTCZYX.forEach((val, idx) => {
+        if (val > -1) {
+          if (val > specLen) {
+            throw new Error("Unexpected axis index");
+          }
+          sliceSpec[val] = unorderedSpec[idx];
+        }
+      });
+
+      const result = (await level.getRaw(sliceSpec)) as RawArray;
+      const u8 = convertChannel(result.data, result.dtype);
+      vol.setChannelDataFromVolume(ch, u8);
+      onChannelLoaded?.(vol, ch);
+    });
+
+    await Promise.all(channelPromises);
```

Comment:
what would happen if you didn't await here?

Reply:
This is just an add-on to make the function return a promise that resolves when all channels are loaded - I don't think this method is ever actually `await`ed. (Come to think of it, this does give us the opportunity to transition this method from callback-based to promise-based with minimal changes, if we want.) As written, I don't believe it'll block anything.